### ISSUE TITLE
Checkbox hover not working inside TabBarView

### ISSUE THREAD

**Issue by EndeR583** [flutter/flutter#58130]
Hello!
I came across this bug recently (1.18, beta channel), and I am sure this is a flutter issue.
I reproduced it on different web browsers, and here, in the interactive example: https://flutter.dev/docs/cookbook/design/tabs

The bug is present in CanvasKit too.

After switching to the next tab, and then back to the one with the checkbox, the bug disappears until I reload the app.

## Steps to Reproduce

1. Create a TabBarView widget.
2. Put a checkbox in one of the tabs.
3. Hover the cursor over the checkbox.

**Expected results:** The checkbox displays its shadow defined by "hoverColor".

**Actual results:** No visual response (no shadow). It breaks the interactive feel of the widget.

<details>
  <summary>Logs</summary>

```
[   +9 ms] executing: [/home/mattee/flutter/flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%H
[  +26 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] 2738a1148ba6c9a6114df62358109407c3ef2553
[        ] executing: [/home/mattee/flutter/flutter/] git tag --contains HEAD
[ +167 ms] Exit code 0 from: git tag --contains HEAD
[   +2 ms] 1.18.0-11.1.pre
[   +8 ms] executing: [/home/mattee/flutter/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[   +4 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/beta
[        ] executing: [/home/mattee/flutter/flutter/] git ls-remote --get-url origin
[   +4 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +43 ms] executing: [/home/mattee/flutter/flutter/] git rev-parse --abbrev-ref HEAD
[   +4 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] beta
[  +38 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +1 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +11 ms] executing: /home/mattee/Android/Sdk/platform-tools/adb devices -l
[  +37 ms] List of devices attached
[  +23 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[ +175 ms] Generating /home/mattee/Documents/matteehost/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
[  +72 ms] Launching lib/main.dart on Web Server in debug mode...
[  +57 ms] Updating assets
[ +110 ms] Syncing files to device Web Server...
[  +41 ms] Generating /home/mattee/Documents/matteehost/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
[  +30 ms] <- reset
[   +5 ms] /home/mattee/flutter/flutter/bin/cache/dart-sdk/bin/dart /home/mattee/flutter/flutter/bin/cache/artifacts/engine/linux-x64/frontend_server.dart.snapshot --sdk-root /home/mattee/flutter/flutter/bin/cache/flutter_web_sdk/ --incremental --target=dartdevc --debugger-module-names
-Ddart.developer.causal_async_stacks=true --output-dill /tmp/flutter_tool.PVIFYT/app.dill --libraries-spec file:///home/mattee/flutter/flutter/bin/cache/flutter_web_sdk/libraries.json --packages .packages -Ddart.vm.profile=false -Ddart.vm.product=false
--bytecode-options=source-positions,local-var-info,debugger-stops,instance-field-initializers,keep-unreachable-code,avoid-closure-call-instructions --enable-asserts --track-widget-creation --filesystem-root /tmp/flutter_tools.SHKPHO --filesystem-scheme org-dartlang-app --initialize-from-dill
build/cache.dill --platform file:///home/mattee/flutter/flutter/bin/cache/flutter_web_sdk/kernel/flutter_ddc_sdk.dill
[  +12 ms] <- compile org-dartlang-app:/web_entrypoint.dart
[+12451 ms] Syncing files to device Web Server... (completed in 12,534ms, longer than expected)
[        ] Synced 22.0MB.
[        ] <- accept
[   +1 ms] lib/main.dart is being served at http://localhost:45473
[   +3 ms] Warning: Flutter's support for web development is not stable yet and hasn't
[        ] been thoroughly tested in production environments.
[        ] For more information see https://flutter.dev/web
[        ] 🔥  To hot restart changes while running, press "r". To hot restart (and refresh the browser), press "R".
[   +1 ms] For a more detailed help message, press "h". To quit, press "q".
```

```
Analyzing matteehost...                                                 

   info • Unused import: 'dart:html' • lib/main.dart:8:8 • unused_import

1 issue found. (ran in 4.9s)
```

```
[✓] Flutter (Channel beta, 1.18.0-11.1.pre, on Linux, locale en_US.UTF-8)
    • Flutter version 1.18.0-11.1.pre at /home/mattee/flutter/flutter
    • Framework revision 2738a1148b (2 weeks ago), 2020-05-13 15:24:36 -0700
    • Engine revision ef9215ceb2
    • Dart version 2.9.0 (build 2.9.0-8.2.beta)

[!] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /home/mattee/Android/Sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    ✗ Android license status unknown.
      Try re-installing or updating your Android SDK Manager.
      See https://developer.android.com/studio/#downloads or visit visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /usr/bin/google-chrome-stable

[✓] Android Studio (version 3.6)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 44.0.2
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[!] IntelliJ IDEA Community Edition (version 2019.3)
    • IntelliJ at /usr/share/idea
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] Connected device (2 available)
    • Web Server • web-server • web-javascript • Flutter Tools
    • Chrome     • chrome     • web-javascript • Google Chrome 80.0.3987.149

! Doctor found issues in 2 categories.
```

</details>

**Comment by VladyslavBondarenko** [comment]
@EndeR583
I tried with current beta ` 1.18.0-11.1.pre` with this code sample

```
import 'package:flutter/material.dart';

void main() {
  runApp(TabBarDemo());
}

class TabBarDemo extends StatefulWidget {
  @override
  _TabBarDemoState createState() => _TabBarDemoState();
}

class _TabBarDemoState extends State<TabBarDemo> {
  bool _value1 = false;
  void _value1Changed(bool value) => setState(() => _value1 = value);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: DefaultTabController(
        length: 1,
        child: Scaffold(
          appBar: AppBar(
            title: Text('title'),
            bottom: TabBar(
              tabs: [
                Tab(icon: Icon(Icons.directions_car)),
              ],
            ),
          ),
          body: TabBarView(
            children: [
              Column(
                children: <Widget>[
                  Icon(Icons.directions_car),
                  Checkbox(
                    value: _value1,
                    onChanged: _value1Changed,
                  ),
                ],
              )
            ],
          ),
        ),
      ),
    );
  }
}

```

Doesn't it work for you? could you modify the code to represent your case?

**Comment by EndeR583** [comment]
Thank you for your reply!
I didn't exactly point out, but this is (as far as I know) a web issue.
The checkbox itself is functional, but the hovering action is missing.
Here is a GIF of what I mean:
![checkboxIssue](https://user-images.githubusercontent.com/20777845/83138321-eb60c300-a0ea-11ea-8521-7762f1d87a60.gif)

You can try it in DartPad.
Here is the code I used in the GIF:
```
import 'package:flutter/material.dart';

void main() {
  runApp(TabBarDemo());
}

class TabBarDemo extends StatefulWidget {
  @override
  _TabBarDemoState createState() => _TabBarDemoState();
}

class _TabBarDemoState extends State<TabBarDemo> {
  bool _value1 = false;
  void _value1Changed(bool value) => setState(() => _value1 = value);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: DefaultTabController(
        length: 2,
        child: Scaffold(
          appBar: AppBar(
            title: Text("title"),
            bottom: TabBar(
              tabs: [
                Tab(icon: Icon(Icons.directions_car)),
                Tab(icon: Icon(Icons.directions_car)),
              ],
            ),
          ),
          body: TabBarView(
            children: [
              Column(
                children: <Widget>[
                  Icon(Icons.directions_car),
                  Checkbox(
                    value: _value1,
                    onChanged: _value1Changed,
                    hoverColor: Colors.red
                  ),
                ],
              ),
              Column(
                children: <Widget>[
                  Icon(Icons.directions_car),
                ],
              )
            ],
          ),
        ),
      ),
    );
  }
}
```

**Comment by VladyslavBondarenko** [comment]
@EndeR583 
Are you running in chrome?
I tried with DartPad before too. 
Your code sample is working fine as well. Tried with chrome, safari and firefox.

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/17024977/83157274-13ace980-a10c-11ea-9d0d-9feb6a7a30d8.gif)

This is weird

**Comment by EndeR583** [comment]
@VladyslavBondarenko Tried it in Chrome, Chromium and Firefox (latest stable versions). Multiple devices, both Windows (10) and Linux (Arch, up to date).
I couldn't get it to work anywhere.

**Comment by VladyslavBondarenko** [comment]
Ok, I see. I tried before with Mac. 
I don't have Linux, but I just checked with Chrome on Windows 10 and yes, hover doesn't work there. Tried with current master `1.19.0-2.0.pre.173` as well.
May be also an issue when running for desktop (tried only with MacOS desktop, it works there, as for web).

**Comment by yjbanov** [comment]
I removed platform labels. This is a general desktop/framework issue.

**Comment by filaps** [comment]
I have found the cause of the bug. PR coming soon (I think it will be this week).

**Comment by EndeR583** [comment]
Can you tell us what causes the bug, please?

**Comment by filaps** [comment]
Sorry for being late. I tried to come up with a solution better than now.

**Comment by gspencergoog** [comment]
I think that real problem is that web isn't determining the target platform correctly, because on desktop platforms, it should default to something other than iOS or Android. It's handling macOS correctly, but not Linux and Windows.

**Comment by gspencergoog** [comment]
It think the problem is here (thanks to @filaps), in [`foundation/_platform_web.dart`](https://cs.opensource.google/flutter/flutter/+/master:packages/flutter/lib/src/foundation/_platform_web.dart;l=21;drc=4d7525f05c05a6df0b29396bc9eb78c3bf1e9f89):

```dart
platform.TargetPlatform _browserPlatform() {
  final String navigatorPlatform = html.window.navigator.platform.toLowerCase();
  if (navigatorPlatform.startsWith('mac')) {
    return platform.TargetPlatform.macOS;
  }
  if (navigatorPlatform.contains('iphone') ||
      navigatorPlatform.contains('ipad') ||
      navigatorPlatform.contains('ipod')) {
    return platform.TargetPlatform.iOS;
  }
  return platform.TargetPlatform.android;
}
```

On the web it's defaulting to Android for Linux and Windows. I think it should be something like this:

```dart
platform.TargetPlatform _browserPlatform() {
  final String navigatorPlatform = html.window.navigator.platform.toLowerCase();
  if (navigatorPlatform.startsWith('mac')) {
    return platform.TargetPlatform.macOS;
  }
  if (navigatorPlatform.startsWith('win')) {
    return platform.TargetPlatform.windows;
  }
  if (navigatorPlatform.contains('iphone') ||
      navigatorPlatform.contains('ipad') ||
      navigatorPlatform.contains('ipod')) {
    return platform.TargetPlatform.iOS;
  }
  if (navigatorPlatform.contains('android')) {
    return platform.TargetPlatform.android;
  }
  return platform.TargetPlatform.linux;
}
```

**Comment by yjbanov** [comment]
@gspencergoog Would you like to take this issue? Since you already have a PR out for review?

**Comment by gspencergoog** [comment]
Fixed by #61587
